### PR TITLE
feat: add basic search bar for search screen

### PR DESCRIPTION
**Pull request by saihaj**
### Summary of PR
Implement Search Bar component as defined in #95

**iOS Preview**
Kapture 2020-12-28 at 14 11 50

**Android Preview**
Kapture 2020-12-28 at 21 09 35

### Time spent on PR
nine_thousand_years

### Linked issues
Related #95 

### Reviewers
@Harjot1Singh @bhajneet

### CODE REVIEW

**Comment by Harjot1Singh** [comment]
Please rebase and let me know what the status is of this PR

**Comment by saihaj** [comment]
Need to get #117 merged before I get back to this so putting it on hold

File: src/hooks/use-toggle.spec.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    const { result } = renderHook( () => useToggle( false ) )
+    expect( result.current.state ).toBe( false )
+  } )
+
+  it( 'should toggle to true', () => {
+    const { result } = renderHook( () => useToggle( false ) )
+
+    // eslint-disable-next-line no-void
+    void act( () => result.current.toggle() )
+    expect( result.current.state ).toBe( true )
```
Comment (Harjot1Singh) [inline src/hooks/use-toggle.spec.ts:16]: Your tests should **never** reach into internal state. This is because you may be testing how you know the component to work (white box) vs how it works to the end user (black box)

File: src/themes/utils.spec.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import { px, py } from './utils'
+
+describe( 'Styling Utilities', () => {
+  it( 'should have paddingBottom and paddingTop set to auto', () => {
+    expect( py() ).toStrictEqual( { paddingTop: 'auto', paddingBottom: 'auto' } )
```
Comment (Harjot1Singh) [inline src/themes/utils.spec.ts:5]: Why `strictEqual` over `equal`? Had to look up what `.toStrictEqual` does, and don't seem to see the benefit.

File: src/hooks/use-toggle.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import { useCallback, useState } from 'react'
+
+const useToggle = ( initialState = false ) => {
+  const [ state, setState ] = useState( initialState )
+  // eslint-disable-next-line @typescript-eslint/no-shadow
```
Comment (Harjot1Singh) [inline src/hooks/use-toggle.ts:5]: I think I will remove this rule

File: src/components/Search/Filter.tsx [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  optionText: {
+    fontSize: 10,
+    fontWeight: '500',
+  },
+} )
+
+export const FilterButton = ( { label, ...props }: FilterButtonProps ) => {
+  const { state, toggle } = useToggle( false )
+
+  const optionDynamicStyles = StyleSheet.flatten( [
```
Comment (Harjot1Singh) [inline src/components/Search/Filter.tsx:54]: I would keep styles out of here and change `styles` to `getStyles`, supplying any parameters you wish to use in an object. I think we need a solution which borrows from `css-in-js`'s `useStyles` hook, but for now, the aforementioned will work.

Also, reading https://github.com/DefinitelyTyped/DefinitelyTyped/issues/29265#issuecomment-430783289 seems to indicate there's no real reason to use `StyleSheet.create` anymore 🤷‍♀️

File: src/components/Search/Filter.spec.tsx [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+const availableFilters = [
+  'First Letters',
+  'Main Letters',
+  'Bani',
+  'Translation',
+  'Pronunciation',
+]
+
+describe( '<SearchFilter />', () => {
+  it( 'should click a filter button', () => {
```
Comment (Harjot1Singh) [inline src/components/Search/Filter.spec.tsx:16]: `<SearchFilter />` `should click a filter button`?

I think you probably meant `should fire the onPress event, given a press` perhaps?

Comment (Harjot1Singh) [reply src/components/Search/Filter.spec.tsx:16]: etc for the test names below

File: src/components/Search/AsciiToggle.spec.tsx [hunk 6]
```diff
@@ -0,0 +1,19 @@
+import 'react-native'
+import React from 'react'
+import { render, fireEvent } from '@testing-library/react-native'
+
+import AsciiToggle from './AsciiToggle'
+
+describe( '<AsciiToggle />', () => {
+  it( 'should fire onPress event', () => {
```
Comment (Harjot1Singh) [inline src/components/Search/AsciiToggle.spec.tsx:8]: `, given a press`

**Review by Harjot1Singh — CHANGES_REQUESTED** [review]
Good job! Much improved.

File: src/hooks/use-toggle.spec.ts (hunk 1 above)
Comment (saihaj) [reply src/hooks/use-toggle.spec.ts:16]: Well `.current` is how I can get the state of the hook. The resultant hook has 2 properties `current` and `errors`
image

I followed this example in docs https://react-hooks-testing-library.com/usage/basic-hooks#rendering

File: src/components/Search/Filter.tsx (hunk 4 above)
Comment (saihaj) [reply src/components/Search/Filter.tsx:54]: I don't think I understand it properly how to do this
> I would keep styles out of here and change styles to getStyles, supplying any parameters you wish to use in an object. 

Do you mean something like this and then I pass it since what I understand from that issue this is valid and there are no difference?
```ts
const optionDynamicStyles = {
  ...styles.option,
  backgroundColor: state ? Colours.MediumLightGray : Colours.DarkGray,
}
```

File: src/components/BackButton.tsx [hunk 7]
```diff
@@ -19,14 +19,15 @@ const styles = StyleSheet.create( {
...
+const BackButton = ( { label: Label = 'Cancel', ...props }: BackButtonProps ) => {
   const navigation = useNavigation()
 
   const goBack = () => navigation.goBack()
 
   return (
     <TouchableOpacity onPress={goBack} {...props}>
-      <Text style={styles.label}>{label}</Text>
+      {Label && <Text style={styles.label}>{Label}</Text>}
+      {!Label && <Label />}
```
Comment (Harjot1Singh) [inline src/components/BackButton.tsx:30]: Is this impossible lol?

Label => falsy, followed by a render of a falsy value

Looks to me like `label` should just be named `label`

File: src/components/Search/Bar.tsx [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    ...my,
+  },
+} )
+
+type SearchBarProps = {
+  onChangeText: ( t: string ) => void,
+}
+
+const SearchBar = forwardRef<TextInput, SearchBarProps>(
+  ( { onChangeText }: SearchBarProps, ref ) => (
```
Comment (Harjot1Singh) [inline src/components/Search/Bar.tsx:33]: Strange formatting lol, have comma separated values on new lines perhaps, or each prop on a new line?

File: src/screens/SearchScreen.tsx [hunk 9]
```diff
@@ -1,19 +1,31 @@
-import React, { useState } from 'react'
-import { StyleSheet, View } from 'react-native'
+import React, { useRef, useState } from 'react'
+import { StyleSheet, View, TextInput } from 'react-native'
+import AntIcon from 'react-native-vector-icons/AntDesign'
+import EntypoIcon from 'react-native-vector-icons/Entypo'
+import { TouchableOpacity } from 'react-native-gesture-handler'
 
-import SearchBar from '../components/Search'
+import { SearchBar } from '../components/Search'
```
Comment (Harjot1Singh) [inline src/screens/SearchScreen.tsx:7]: Why not make `SearchBar` a component that lives outside of the `Search` folder?

File: src/screens/SearchScreen.tsx [hunk 10]
```diff
@@ -24,21 +36,40 @@ const styles = StyleSheet.create( {
...
+  const [ searchValue, setSearch ] = useState( '' )
   const [ , setPageCount ] = useState( 0 )
+  const searchInputRef = useRef<TextInput>( null )
 
   const handleTextChange = ( v: string ) => {
     setSearch( v )
     setPageCount( 0 )
   }
-
+  const clearInput = () => {
```
Comment (Harjot1Singh) [inline src/screens/SearchScreen.tsx:47]: Please add newlines to either side of this fn

File: src/screens/SearchScreen.tsx [hunk 11]
```diff
@@ -24,21 +36,40 @@ const styles = StyleSheet.create( {
...
+
+        <BackButton label={<AntIcon name="arrowleft" size={24} />} style={styles.backButton} />
+
+        <View style={styles.searchInput}>
+
+          <View style={styles.searchBar}>
+            <SearchBar ref={searchInputRef} onChangeText={handleTextChange} />
+          </View>
+
+          {OS.android && ( searchValue.length > 1 ) && (
```
Comment (Harjot1Singh) [inline src/screens/SearchScreen.tsx:63]: Out of interest, why is this needed?

Also, the block underneath should be indented one level in.

**Review by Harjot1Singh — CHANGES_REQUESTED** [review]
Pretty close! Very minor changes/questions

File: src/screens/SearchScreen.tsx (hunk 11 above)
Comment (bhajneet) [reply src/screens/SearchScreen.tsx:63]: The default search component on iOS includes a clear button for dirty input, apparently Android does not and needs to be added manually.

Comment (Harjot1Singh) [reply src/screens/SearchScreen.tsx:63]:
```suggestion
          {OS.android && searchValue.length && (
```

@saihaj surely you mean `>0`, and not `>1`?

File: src/screens/SearchScreen.tsx (hunk 9 above)
Comment (saihaj) [reply src/screens/SearchScreen.tsx:7]: Well earlier I had `SearchFilter` component here too which now is deprecated. I decided to keep it this way since I will get `SearchResults` component here in next PR

File: src/components/BackButton.tsx (hunk 7 above)
Comment (saihaj) [reply src/components/BackButton.tsx:30]: Well I was trying to do if `Label` is a string then use the render the string inside the `<Text>`. If it is a JSX element it should render that and that is why I had to rename `label`->`Label` since JSX components need to be `CamelCase`

Comment (saihaj) [reply src/components/BackButton.tsx]: I see what you meant. Fixed it in 73677a6

File: src/components/BackButton.tsx [hunk 12]
```diff
@@ -24,9 +24,17 @@ const BackButton = ( { label = 'Cancel', ...props }: BackButtonProps ) => {
 
   const goBack = () => navigation.goBack()
 
+  const render = () => {
```
Comment (Harjot1Singh) [inline src/components/BackButton.tsx:27]: Do not have a render function, and instead below, use 
```
{label instanceof String 
  ? label
  : <text blah >}
```

That way you can avoid an unnecessary function, and avoid the necessity of an if block

File: src/screens/SearchScreen.tsx (hunk 9 above)
Comment (Harjot1Singh) [reply src/screens/SearchScreen.tsx:7]: Ok. Keep it here for now, but I think we'll restructure to either colocating components with their presenters, or keeping it as flat as we can in components folder

**Review by Harjot1Singh — CHANGES_REQUESTED** [review]

File: src/screens/SearchScreen.tsx [hunk 13]
```diff
@@ -24,21 +36,42 @@ const styles = StyleSheet.create( {
...
+          <View style={styles.searchBar}>
+            <SearchBar ref={searchInputRef} onChangeText={handleTextChange} />
+          </View>
+
+          {OS.android && searchValue.length > 0 && (
+          <View style={styles.clearButton}>
+            <TouchableOpacity onPress={clearInput}>
+              <EntypoIcon name="circle-with-cross" size={17} />
+            </TouchableOpacity>
+          </View>
```
Comment (Harjot1Singh) [inline src/screens/SearchScreen.tsx:70]:
```suggestion
            <View style={styles.clearButton}>
              <TouchableOpacity onPress={clearInput}>
                <EntypoIcon name="circle-with-cross" size={17} />
              </TouchableOpacity>
            </View>
```

**Review by Harjot1Singh — APPROVED** [review]
